hipcheck/src/analysis/result.rs

```diff
@@ -0,0 +1,136 @@
+use crate::hc_error;
+use crate::report::Concern;
+use crate::Result;
+use crate::F64;
+
+/// Represents the enhanced result of a hipcheck analysis. Contains the actual outcome
+/// of the analysis, plus additional meta-information the analysis wants to provide to
+/// HipCheck core, such as raised concerns.
+#[allow(dead_code)]
+pub struct HCAnalysisResult {
+	pub outcome: AnalysisOutcome,
+	pub concerns: Vec<Concern>,
+}
+
+#[allow(dead_code)]
+pub type SkippableAnalysisResult = Option<HCAnalysisResult>;
+
+/// Represents the result of a hipcheck analysis. Either the analysis encountered
+/// an error, or it completed and returned a value.
+#[allow(dead_code)]
+pub enum AnalysisOutcome {
+	Error(HCAnalysisError),
+	Completed(HCAnalysisValue),
+}
+
+/// Enumeration of potential errors that a HipCheck analysis might return. The Generic
+/// variant enables representing errors that aren't covered by other variants.
+#[allow(dead_code)]
+pub enum HCAnalysisError {
+	Generic(crate::error::Error),
+}
+
+/// A HipCheck analysis may return a basic or composite value. By splitting the types
+/// into two sub-enums under this one, we can eschew a recursive enum definition and
+/// ensure composite types only have a depth of one.
+#[allow(dead_code)]
+pub enum HCAnalysisValue {
+	Basic(HCBasicValue),
+	Composite(HCCompositeValue),
+}
+
+/// Basic HipCheck analysis return types
+#[derive(Debug)]
+pub enum HCBasicValue {
+	Integer(i64),
+	Unsigned(u64),
+	Float(F64),
+	Bool(bool),
+	String(String),
+}
+impl From<i64> for HCBasicValue {
+	fn from(value: i64) -> Self {
+		HCBasicValue::Integer(value)
+	}
+}
+impl From<u64> for HCBasicValue {
+	fn from(value: u64) -> Self {
+		HCBasicValue::Unsigned(value)
+	}
+}
+impl From<F64> for HCBasicValue {
+	fn from(value: F64) -> Self {
+		HCBasicValue::Float(value)
+	}
+}
+impl TryFrom<f64> for HCBasicValue {
+	type Error = crate::Error;
+	fn try_from(value: f64) -> Result<HCBasicValue> {
+		let inner = F64::new(value)?;
+		Ok(HCBasicValue::Float(inner))
+	}
+}
+impl From<bool> for HCBasicValue {
+	fn from(value: bool) -> Self {
+		HCBasicValue::Bool(value)
+	}
+}
+impl From<String> for HCBasicValue {
+	fn from(value: String) -> Self {
+		HCBasicValue::String(value)
+	}
+}
+impl From<&str> for HCBasicValue {
```

Comment:
Could probably consolidate the `From<&str>` and `From<String>` impls with `impl<T: AsRef<str>> From<T> for HCBasicValue`. Right now `From<String>` will take ownership of `String` rather than copying or taking by reference the way it does for everything else, which means callers are likely to need to do their own `clone` anyway, so an `as_ref().to_owned()` chain isn't that big a deal (this is basically a clone).

Reply:
When I tried to do this, I got the attached "conflicting implementations" error message. So instaed I just opted to delete the `From<String>` impl, and converters from `String` can just call `as_str()` first.
<img width="840" alt="conflicting_impls_error" src="https://github.com/mitre/hipcheck/assets/110619939/ed5d8ad5-2483-4b97-b003-b2abb75c4186">

Reply:
Ah, overlapping impls. Classic. Sounds like a good solution!